### ISSUE TITLE
create_expand_pow_optimization doesn’t respect sums

### ISSUE THREAD

**Issue by Wrzlprmft** [sympy/sympy#20753]
Consider the following example:

``` Python 3
from sympy.printing import ccode
from sympy.abc import a, b
from sympy.codegen.rewriting import create_expand_pow_optimization

expand_opt = create_expand_pow_optimization(3)

print(ccode(expand_opt((a+b)**2))
```

This yields `pow(a + b, 2)`, while I would expect it to yield `(a+b)*(a+b)`.

**Comment by Shardul555** [comment]
The requirements for expansions are that the base needs to be a symbol and the exponent needs to be an Integer (and be less than or equal to limit). 
I think as the base is an expression hence expansion not occured.

**Comment by bjodah** [comment]
Thank you @Wrzlprmft for opening this issue. I opened a PR with my suggestion of how to support this use case. Any feedback of yours is most welcome over there (gh-20788).

**Comment by oscarbenjamin** [comment]
I would have expected that modern C compilers can already perform these kinds of optimisations in both directions as needed.

**Comment by Wrzlprmft** [comment]
They usually can, but depending on the situation, optimisation takes a long time, this is the only one that really makes a difference for you, and the compiler doesn’t make it easy to separately toggle this.

**Comment by oscarbenjamin** [comment]
If the compiler doesn't optimise `pow(a+b, 2)` then would you expect it to be able to optimise the duplicate expressions in `(a+b)*(a+b)`?

**Comment by Wrzlprmft** [comment]
> If the compiler doesn't optimise `pow(a+b, 2)` then would you expect it to be able to optimise the duplicate expressions in `(a+b)*(a+b)`?

Given the circumstances, yes. Also not optimising would usually still be the lesser evil. For simplicity, let’s consider a **literal** implementation of the three following options:

1.  ``` C
     double c = a+b;
     double x = pow(c,2);
    ```
2.  ``` C
     double c = a+b;
     double x = c*c;
    ```
3.  ``` C
     double c = a+b;
     double d = a+b;
     double x = c*d;
    ```

Option 1 uses a series expansion to evaluate `pow`, which in turn features many multiplications and additions. It is therefore the slowest option (unless the computation of `c` is very complex). Also due to the series expansion, the result cannot be guaranteed to be the same as for Options 2 and 3. By contrast Options 2 and 3 will always produce the same result. Therefore Option 1 will only be optimised to Option 2 when you allow optimisations that do not exactly preserve the result.

**Comment by oscarbenjamin** [comment]
> Therefore Option 1 will only be optimised to Option 2 when you allow optimisations that do not exactly preserve the result.

Oh that's a good point.

**Comment by bjodah** [comment]
Fixed in #20788
